File: go/trees/0226_invert_binary_tree/invert_binary_tree_test.go
package invertbinarytree

import (
	"math"
	"reflect"
	"testing"
)

const NULL = math.MinInt32

func buildTree(nums []int) *TreeNode {
	if len(nums) == 0 || nums[0] == NULL {
		return nil
	}

	root := &TreeNode{Val: nums[0]}
	queue := []*TreeNode{root}
	i := 1

	for i < len(nums) {
		curr := queue[0]
		queue = queue[1:]

		if i < len(nums) && nums[i] != NULL {
			curr.Left = &TreeNode{Val: nums[i]}
			queue = append(queue, curr.Left)
		}
		i++

		if i < len(nums) && nums[i] != NULL {
			curr.Right = &TreeNode{Val: nums[i]}
			queue = append(queue, curr.Right)
		}
		i++
	}
	return root
}

func TestInvertTree(t *testing.T) {
	tests := []struct {
		name     string
		input    []int
		expected []int
	}{
		{
			name:     "Example 1 - Full Tree",
			input:    []int{4, 2, 7, 1, 3, 6, 9},
			expected: []int{4, 7, 2, 9, 6, 3, 1},
		},
		{
			name:     "Example 2 - Small Tree",
			input:    []int{2, 1, 3},
			expected: []int{2, 3, 1},
		},
		{
			name:     "Example 3 - Empty Tree",
			input:    []int{},
			expected: []int{},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			root := buildTree(tt.input)
			expectedTree := buildTree(tt.expected)

			result := InvertTree(root)

			if !reflect.DeepEqual(result, expectedTree) {
				t.Errorf("got %v, want %v", result, expectedTree)
			}
		})
	}
}

func BenchmarkInvertTreeRecursive(b *testing.B) {
	input := []int{4, 2, 7, 1, 3, 6, 9, 10, 11, 12, 13, 14, 15, 16, 17}
	root := buildTree(input)

	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		InvertTree(root)
	}
}

File: c/trees/0226_invert_binary_tree/invert_binary_tree_test.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

#include "invert_binary_tree.c"

#define RUN_TEST(name, condition)    \
    if (condition)                   \
    {                                \
        printf("[PASS] %s\n", name); \
    }                                \
    else                             \
    {                                \
        printf("[FAIL] %s\n", name); \
    }

struct TreeNode *createNode(int val)
{
    struct TreeNode *node = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    node->val = val;
    node->left = NULL;
    node->right = NULL;
    return node;
}

void freeTree(struct TreeNode *root)
{
    if (root == NULL)
        return;
    freeTree(root->left);
    freeTree(root->right);
    free(root);
}

bool compareTrees(struct TreeNode *t1, struct TreeNode *t2)
{
    if (t1 == NULL && t2 == NULL)
        return true;
    if (t1 == NULL || t2 == NULL)
        return false;
    if (t1->val != t2->val)
        return false;

    return compareTrees(t1->left, t2->left) && compareTrees(t1->right, t2->right);
}

void TestInvertTree()
{
    printf("--- RUNNING TESTS ---\n");

    struct TreeNode *t1 = NULL;
    struct TreeNode *expected1 = NULL;
    struct TreeNode *got1 = invertTree(t1);
    RUN_TEST("Empty tree", compareTrees(got1, expected1));
    freeTree(got1);
    freeTree(expected1);

    struct TreeNode *t2 = createNode(1);
    struct TreeNode *expected2 = createNode(1);
    struct TreeNode *got2 = invertTree(t2);
    RUN_TEST("Single node", compareTrees(got2, expected2));
    freeTree(got2);
    freeTree(expected2);

    struct TreeNode *t3 = createNode(1);
    t3->left = createNode(2);

    struct TreeNode *expected_t3 = createNode(1);
    expected_t3->right = createNode(2);

    struct TreeNode *got3 = invertTree(t3);
    RUN_TEST("Left-heavy tree", compareTrees(got3, expected_t3));
    freeTree(got3);
    freeTree(expected_t3);

    struct TreeNode *t4 = createNode(4);
    t4->left = createNode(2);
    t4->right = createNode(7);
    t4->left->left = createNode(1);
    t4->left->right = createNode(3);
    t4->right->left = createNode(6);
    t4->right->right = createNode(9);

    struct TreeNode *expected_t4 = createNode(4);
    expected_t4->left = createNode(7);
    expected_t4->right = createNode(2);
    expected_t4->left->left = createNode(9);
    expected_t4->left->right = createNode(6);
    expected_t4->right->left = createNode(3);
    expected_t4->right->right = createNode(1);

    struct TreeNode *got4 = invertTree(t4);
    RUN_TEST("Full binary tree", compareTrees(got4, expected_t4));
    freeTree(got4);
    freeTree(expected_t4);
}

struct TreeNode *createBalancedTree(int depth)
{
    if (depth == 0)
        return NULL;
    struct TreeNode *node = createNode(depth);
    node->left = createBalancedTree(depth - 1);
    node->right = createBalancedTree(depth - 1);
    return node;
}

void BenchmarkInvertTree()
{
    printf("\n--- RUNNING BENCHMARK ---\n");

    int iterations = 10000;
    int depth = 10;

    clock_t start_time = clock();

    for (int i = 0; i < iterations; i++)
    {
        struct TreeNode *root = createBalancedTree(depth);
        struct TreeNode *inverted = invertTree(root);

        freeTree(inverted);
    }

    clock_t end_time = clock();
    double time_taken = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

    printf("BenchmarkInvertTree\t%d iteracoes\t%.4f segundos totais\n", iterations, time_taken);
}

int main()
{
    TestInvertTree();
    BenchmarkInvertTree();

    return 0;
}